File: week3/peak-index-in-a-mountain-array.java
class Solution {
    public int peakIndexInMountainArray(int[] A) {
        return peakSearch(A, 0, A.length - 1);
    }
    
    public int peakSearch(int[] A, int lo, int hi) {
        if (lo >= hi) {
            return lo;
        }
        int mid = lo + (hi - lo) / 2;
        // We are on the ascending side of the mountain
        if (A[mid] < A[mid + 1]) {
            return peakSearch(A, mid + 1, hi);
        // We are on the descending side of the mountain
        } else {
            return peakSearch(A, lo, mid);
        }
    }
}

File: 2026/week3/peak-index-in-a-mountain-array.py
class Solution:
    def peakIndexInMountainArray(self, arr: List[int]) -> int:
        lo = 0
        hi = len(arr) - 1
        while lo < hi:
            mid = lo + (hi - lo) // 2
            if arr[mid] < arr[mid + 1]:
                lo = mid + 1
            else:
                hi = mid
        return lo
